### ISSUE TITLE
Question: Observable stuck in case StackoverflowException thrown inside Task {}, is it expected?

### ISSUE THREAD

**Issue by GrigorievNick** [monix/monix#1671]
```
import monix.eval.Task
import monix.execution.Scheduler.Implicits.global
import monix.reactive.Consumer
import monix.reactive.Observable
import org.scalatest.FunSuite

import scala.concurrent.duration._

class MonixExceptionTest extends FunSuite {

  test("fatal") {
    Observable
      .fromIterable(0 to 10)
      .mapParallelUnordered(4)(i =>
        Task {
          try throw new StackOverflowError(s"$i")
          catch {
//            case x: Throwable => throw new RuntimeException(x)
            case x: Throwable => throw x
          }
          i
        }
      )
      .consumeWith(Consumer.foreach(println))
      .runSyncUnsafe()
  }

}
```

this code will be stuck forever. \
But if we wrap Stackoverflow with RuntimeException, Observable will fail.
Do this expected behave?

**Comment by ghik** [comment]
Yes, this is because `StackOverflowError` is considered fatal and unrecoverable. 

Traditionally, all `Error`s are considered fatal (as opposed to `Exception`s). Scala has a slightly refined definition of what is _fatal_ (see `scala.util.control.NonFatal` extractor) but `StackOverflowError` falls into both definitions so Monix will not try to catch and recover from it.

**Comment by GrigorievNick** [comment]
OK, but still  I expect to fail observable, rather then just  silently stuck.
Or I miss understand something?

**Comment by ghik** [comment]
For an `Observable` to fail on `StackOverflowError`, it would have to catch it at some point. But fatal errors should never be caught. They should be allowed to crash the process.

Although I remember that at some point `NonFatal` did not recognize `StackOverflowError` as fatal so it may be somewhere in a gray area.

**Comment by GrigorievNick** [comment]
I am not sure that problem on observable side, look like Task is never complete with fail or success - no matter.

**Comment by ghik** [comment]
Yes, but `Task` doesn't catch it for exactly the same reason.

**Comment by GrigorievNick** [comment]
OK, so what is  way to handle this improve error handling on Monix side, or always wrapp code inside `Task ` with try-catch, or add timeout to every `Task`?

**Comment by GrigorievNick** [comment]
Do you happen to have any news on this?
Code must throw an exception and not be silently stuck even without logs.
